storage/storage.c

```diff
@@ -472,79 +465,155 @@ static secbool is_not_wipe_code(const uint8_t *pin, size_t pin_len) {
-                       const uint8_t *random_salt, const uint8_t *ext_salt,
+                       const uint8_t *storage_salt, const uint8_t *ext_salt,
                        uint8_t kek[SHA256_DIGEST_LENGTH],
                        uint8_t keiv[SHA256_DIGEST_LENGTH]) {
-  uint8_t salt[HARDWARE_SALT_SIZE + RANDOM_SALT_SIZE + EXTERNAL_SALT_SIZE] = {
+  uint8_t salt[HARDWARE_SALT_SIZE + STORAGE_SALT_SIZE + EXTERNAL_SALT_SIZE] = {
       0};
   size_t salt_len = 0;
 
   memcpy(salt + salt_len, hardware_salt, HARDWARE_SALT_SIZE);
   salt_len += HARDWARE_SALT_SIZE;
 
-  memcpy(salt + salt_len, random_salt, RANDOM_SALT_SIZE);
-  salt_len += RANDOM_SALT_SIZE;
+  memcpy(salt + salt_len, storage_salt, STORAGE_SALT_SIZE);
+  salt_len += STORAGE_SALT_SIZE;
 
   if (ext_salt != NULL) {
     memcpy(salt + salt_len, ext_salt, EXTERNAL_SALT_SIZE);
     salt_len += EXTERNAL_SALT_SIZE;
   }
 
-  uint32_t progress = (ui_total - ui_rem) * 1000 / ui_total;
-  if (ui_callback && ui_message) {
-    ui_callback(ui_rem, progress, ui_message);
-  }
+  ui_progress(0);
 
   PBKDF2_HMAC_SHA256_CTX ctx = {0};
   pbkdf2_hmac_sha256_Init(&ctx, pin, pin_len, salt, salt_len, 1);
   for (int i = 1; i <= 5; i++) {
     pbkdf2_hmac_sha256_Update(&ctx, PIN_ITER_COUNT / 10);
-    if (ui_callback && ui_message) {
-      progress =
-          ((ui_total - ui_rem) * 1000 + i * DERIVE_SECS * 100) / ui_total;
-      ui_callback(ui_rem - i * DERIVE_SECS / 10, progress, ui_message);
-    }
+    ui_progress(i * PIN_DERIVE_SECS * 100);
   }
   pbkdf2_hmac_sha256_Final(&ctx, kek);
 
   pbkdf2_hmac_sha256_Init(&ctx, pin, pin_len, salt, salt_len, 2);
   for (int i = 6; i <= 10; i++) {
     pbkdf2_hmac_sha256_Update(&ctx, PIN_ITER_COUNT / 10);
-    if (ui_callback && ui_message) {
-      progress =
-          ((ui_total - ui_rem) * 1000 + i * DERIVE_SECS * 100) / ui_total;
-      ui_callback(ui_rem - i * DERIVE_SECS / 10, progress, ui_message);
-    }
+    ui_progress(i * PIN_DERIVE_SECS * 100);
   }
   pbkdf2_hmac_sha256_Final(&ctx, keiv);
 
-  ui_rem -= DERIVE_SECS;
+  ui_rem -= PIN_DERIVE_SECS;
   memzero(&ctx, sizeof(PBKDF2_HMAC_SHA256_CTX));
   memzero(&salt, sizeof(salt));
 }
+#endif
+
+#if USE_OPTIGA
+static void derive_kek_optiga(
+    const uint8_t optiga_secret[OPTIGA_PIN_SECRET_SIZE],
+    const uint8_t *ext_salt, uint8_t kek[SHA256_DIGEST_LENGTH],
+    uint8_t keiv[SHA256_DIGEST_LENGTH]) {
+  uint8_t salt[HARDWARE_SALT_SIZE + EXTERNAL_SALT_SIZE] = {0};
+  size_t salt_len = 0;
+
+  memcpy(salt + salt_len, hardware_salt, HARDWARE_SALT_SIZE);
+  salt_len += HARDWARE_SALT_SIZE;
+
+  if (ext_salt != NULL) {
+    memcpy(salt + salt_len, ext_salt, EXTERNAL_SALT_SIZE);
+    salt_len += EXTERNAL_SALT_SIZE;
+  }
+
+  PBKDF2_HMAC_SHA256_CTX ctx = {0};
+  pbkdf2_hmac_sha256_Init(&ctx, optiga_secret, OPTIGA_PIN_SECRET_SIZE, salt,
+                          salt_len, 1);
+  // TODO Do we even need this one iteration or can we use 0 or completely leave
+  // out pbkdf2_hmac_sha256_Update()?
```

Comment:
I haven't had time to look into this yet. Opinions @onvej-sl?

Reply:
Let's leave the code as is. One iteration takes place automatically in the Init function and `pbkdf2_hmac_sha256_Update(&ctx, 1);` is a actually a nop, but it expresses that one iteration has taken place, which is useful for documentation purposes.